src/mito2/src/memtable/bulk/row_group_reader.rs

```diff
@@ -0,0 +1,223 @@
+use crate::sst::parquet::row_group::{ColumnChunkIterator, RowGroupBase};
+use crate::sst::parquet::DEFAULT_READ_BATCH_SIZE;
+
+/// Helper for reading specific row group inside Memtable Parquet parts.
+// This is similar to [mito2::sst::parquet::row_group::InMemoryRowGroup] since
+// it's a workaround for lacking of keyword generics.
+pub struct MemtableRowGroupPageFetcher<'a> {
+    /// Shared structs for reading row group.
+    base: RowGroupBase<'a>,
+    bytes: Bytes,
+}
+
+impl<'a> MemtableRowGroupPageFetcher<'a> {
+    pub(crate) fn create(
+        row_group_idx: usize,
+        parquet_meta: &'a ParquetMetaData,
+        bytes: Bytes,
+    ) -> Self {
+        let metadata = parquet_meta.row_group(row_group_idx);
+        let row_count = metadata.num_rows() as usize;
+        let page_locations = parquet_meta
+            .offset_index()
+            .map(|x| x[row_group_idx].as_slice());
+
+        Self {
+            base: RowGroupBase {
+                metadata,
+                page_locations,
+                row_count,
+                column_chunks: vec![None; metadata.columns().len()],
+                column_uncompressed_pages: vec![None; metadata.columns().len()],
+            },
+            bytes,
+        }
+    }
+
+    /// Fetches column pages from memory file.
+    pub(crate) fn fetch(&mut self, projection: &ProjectionMask, selection: Option<&RowSelection>) {
+        if let Some((selection, page_locations)) = selection.zip(self.base.page_locations) {
+            // Selection provided.
+            let (fetch_ranges, page_start_offsets) =
+                self.base
+                    .calc_sparse_read_ranges(projection, page_locations, selection);
+            if fetch_ranges.is_empty() {
+                return;
+            }
+            let chunk_data = self.fetch_bytes(&fetch_ranges);
+
+            self.base
+                .assign_sparse_chunk(projection, chunk_data, page_start_offsets);
+        } else {
+            let fetch_ranges = self.base.calc_dense_read_ranges(projection);
+            if fetch_ranges.is_empty() {
+                // Nothing to fetch.
+                return;
+            }
+            let chunk_data = self.fetch_bytes(&fetch_ranges);
+            self.base.assign_dense_chunk(projection, chunk_data);
+        }
+    }
+
+    fn fetch_bytes(&self, ranges: &[Range<u64>]) -> Vec<Bytes> {
+        ranges
+            .iter()
+            .map(|range| self.bytes.slice(range.start as usize..range.end as usize))
+            .collect()
+    }
+
+    /// Creates a page reader to read column at `i`.
+    fn column_page_reader(&self, i: usize) -> parquet::errors::Result<Box<dyn PageReader>> {
+        if let Some(cached_pages) = &self.base.column_uncompressed_pages[i] {
+            debug_assert!(!cached_pages.row_group.is_empty());
+            // Hits the row group level page cache.
+            return Ok(Box::new(RowGroupCachedReader::new(&cached_pages.row_group)));
+        }
```

Comment:
Maybe we don't need to cache uncompressed pages for memtables.

Reply:
yes. we can skip the check here. 0d0d6340f1